### PR TITLE
fix coverity defects with cid (1509)21-24

### PR DESCRIPTION
**Pull request by luozhengzheng**
coverity defects:
CID 150921: Unchecked return value (CHECKED_RETURN)
CID 150924 : Unchecked return value (CHECKED_RETURN)

thanks.

### CODE REVIEW

File: module/zfs/arc.c [hunk 1]
```diff
@@ -3472,7 +3472,8 @@ arc_prune_async(int64_t adjust)
 
 		refcount_add(&ap->p_refcnt, ap->p_pfunc);
 		ap->p_adjust = adjust;
-		taskq_dispatch(arc_prune_taskq, arc_prune_task, ap, TQ_SLEEP);
+		(void) taskq_dispatch(arc_prune_taskq, arc_prune_task,
+			ap, TQ_SLEEP);
```
Comment (behlendorf) [inline module/zfs/arc.c]: This is actually a real potential bug.  In the failure case where `taskq_dispatch()` fails we need to drop the `p_refcnt` which would normally have been run by `arc_prune_task` but now won't be.

File: module/zfs/zfs_vnops.c [hunk 2]
```diff
@@ -941,7 +941,7 @@ zfs_iput_async(struct inode *ip)
 	ASSERT(os != NULL);
 
 	if (atomic_read(&ip->i_count) == 1)
-		taskq_dispatch(dsl_pool_iput_taskq(dmu_objset_pool(os)),
+		(void) taskq_dispatch(dsl_pool_iput_taskq(dmu_objset_pool(os)),
```
Comment (behlendorf) [inline module/zfs/zfs_vnops.c]: This is also a real potential issue.  In the case where the dispatch fails we don't have any safe option except retrying the dispatch and hoping for the best.  We can't call `iput()` synchronously because it's possible it will re-enter the vfs and result in a deadlock.  This is why we have an async version in the first place.  And we can't ignore the error because we'll leak a reference and no longer be able to unmount the filesystem.

The good news is that this should be completely impossible given the current taskq_dispatch implementation and flags passed.  I'd suggest the dropping this hunk from this PR and opening a new one with this alternate untested change.


```diff
diff --git a/module/zfs/zfs_vnops.c b/module/zfs/zfs_vnops.c
index 887b7e1..9ac1a3f 100644
--- a/module/zfs/zfs_vnops.c
+++ b/module/zfs/zfs_vnops.c
@@ -932,19 +932,31 @@ zfs_write(struct inode *ip, uio_t *uio, int ioflag, cred_t
 }
 EXPORT_SYMBOL(zfs_write);
 
+/*
+ * Drop a reference on the passed inode asynchronously.  This ensures
+ * that the caller will never drop the last reference on an inode in
+ * the current context.  Doing so while holding open a tx could result
+ * in a deadlock if iput_final() re-enters the filesystem code.
+ */
 void
 zfs_iput_async(struct inode *ip)
 {
        objset_t *os = ITOZSB(ip)->z_os;
+       taskqid_t id = 0;
+
 
        ASSERT(atomic_read(&ip->i_count) > 0);
        ASSERT(os != NULL);
 
-       if (atomic_read(&ip->i_count) == 1)
-               taskq_dispatch(dsl_pool_iput_taskq(dmu_objset_pool(os)),
-                   (task_func_t *)iput, ip, TQ_SLEEP);
-       else
+       if (atomic_read(&ip->i_count) == 1) {
+               while (id == 0) {
+                       id = taskq_dispatch(
+                           dsl_pool_iput_taskq(dmu_objset_pool(os)),
+                           (task_func_t *)iput, ip, TQ_SLEEP);
+               }
+       } else {
                iput(ip);
+       }
 }
 
 void
```

Comment (luozhengzheng) [inline module/zfs/zfs_vnops.c]: @behlendorf  I will follow your advice.thanks.

File: module/zfs/arc.c [hunk 3]
```diff
@@ -3472,7 +3473,12 @@ arc_prune_async(int64_t adjust)
 
 		refcount_add(&ap->p_refcnt, ap->p_pfunc);
 		ap->p_adjust = adjust;
-		taskq_dispatch(arc_prune_taskq, arc_prune_task, ap, TQ_SLEEP);
+		id = taskq_dispatch(arc_prune_taskq, arc_prune_task,
+			ap, TQ_SLEEP);
+		if (id == 0) {
+			refcount_remove(&ap->p_refcnt, ap->p_pfunc);
+			return;
+		}
```
Comment (behlendorf) [inline module/zfs/arc.c]: In this case you need to `continue` and not return.  It's entirely possible the dispatch will succeed for other entries in the list.  And you need to make sure you bump the counters and drop the mutex.  Something like this.

```
                if ((id = taskq_dispatch(arc_prune_taskq, arc_prune_task, 
                    ap, TQ_SLEEP)) == 0)
                        refcount_remove(&ap->p_refcnt, ap->p_pfunc);
```

Again, it probably best to break this in to its own PR for review/comment so it doesn't hold up these other trivial fixes.

Comment (luozhengzheng) [inline module/zfs/arc.c]: @behlendorf  ok, I will pull a new PR for it. thanks.
